code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/ServiceActivate.java

```diff
@@ -0,0 +1,235 @@
+import io.cattle.platform.core.dao.GenericMapDao;
+import io.cattle.platform.core.dao.GenericResourceDao;
+import io.cattle.platform.core.model.Environment;
+import io.cattle.platform.core.model.Image;
+import io.cattle.platform.core.model.Instance;
+import io.cattle.platform.core.model.Service;
+import io.cattle.platform.core.model.ServiceExposeMap;
+import io.cattle.platform.deferred.util.DeferredUtils;
+import io.cattle.platform.engine.handler.HandlerResult;
+import io.cattle.platform.engine.process.ProcessInstance;
+import io.cattle.platform.engine.process.ProcessState;
+import io.cattle.platform.engine.process.impl.ProcessCancelException;
+import io.cattle.platform.json.JsonMapper;
+import io.cattle.platform.object.ObjectManager;
+import io.cattle.platform.object.process.StandardProcess;
+import io.cattle.platform.object.resource.ResourceMonitor;
+import io.cattle.platform.object.resource.ResourcePredicate;
+import io.cattle.platform.object.util.DataAccessor;
+import io.cattle.platform.object.util.DataUtils;
+import io.cattle.platform.process.common.handler.AbstractObjectProcessHandler;
+import io.cattle.platform.process.progress.ProcessProgress;
+import io.cattle.platform.servicediscovery.api.constants.ServiceDiscoveryConstants;
+import io.cattle.platform.servicediscovery.resource.ServiceDiscoveryConfigItem;
+import io.cattle.platform.servicediscovery.service.ServiceDiscoveryService;
+import io.cattle.platform.storage.service.StorageService;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+@Named
+public class ServiceActivate extends AbstractObjectProcessHandler {
+    @Inject
+    JsonMapper jsonMapper;
+
+    @Inject
+    StorageService storageService;
+
+    @Inject
+    ObjectManager objectManager;
+
+    @Inject
+    GenericMapDao mapDao;
+    
+    @Inject
+    ResourceMonitor resourceMonitor;
+
+    @Inject
+    GenericResourceDao resourceDao;
+
+    @Inject
+    ProcessProgress progress;
+
+    @Inject
+    ServiceDiscoveryService sdServer;
+
+    @Override
+    public String[] getProcessNames() {
+        return new String[] { ServiceDiscoveryConstants.PROCESS_SERVICE_ACTIVATE };
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        Service service = (Service) state.getResource();
+        List<Integer> consumedServiceIds = new ArrayList<>();
+        boolean activateConsumedServices = DataAccessor.fromMap(state.getData()).withScope(ServiceActivate.class)
+                .withKey(ServiceDiscoveryConstants.FIELD_ACTIVATE_CONSUMED_SERVICES).withDefault(false)
+                .as(Boolean.class);
+        if (activateConsumedServices) {
+            // the order is determined only based on volumes_from (service) attribute
```

Comment:
@cjellick Service launch order within the environment, doesn't matter in most of the cases as upon service activation we do DNS configuration as a late binding. Although if service1 uses service2's volumes (volumes_from parameter in docker-compose.yml), its important that the service2 is activated first. Parameter FIELD_ACTIVATE_CONSUMED_SERVICES in the handler indicates that this parameter has been set, and consumed services for the target service have to be activated first. Although I don't specifically extract the consumed service referenced via volumes_from; I just activate all the consumed services - might revise it later.